### PR TITLE
Add `path` as an `einsum` python API kwarg

### PR DESCRIPTION
**Pull request by janeyx99**
Adds an `optimize` kwarg to `torch.einsum` to give users more control. We've attempted to match numpy's API for `optimize` reasonably by discussing with Ralf from numpy.

The kwarg can be:
- a list => we will use this list immediately instead of trying to compute an optimal one
- 'skip' => we will skip the computation and use the input order
- 'auto' => we will attempt to compute the optimal path however we can. Currently, we use opt_einsum and thus will **error** if opt_einsum cannot be found. 
- 'optimal' and 'greedy' => we will attempt to compute the optimal path through the corresponding 'optimal' and 'greedy' heuristics on opt_einsum and will **error** if opt_einsum cannot be found. 
- None => the default is to use opt_einsum if it's available in the env + skip without failing otherwise.

Not having None caused backwards compatibility with vision which may rely on older compiled einsum code, causing failures like https://github.com/pytorch/pytorch/actions/runs/3162886037/jobs/5149977472

Test plan:
CI

I've locally verified that
- it will error if you pass in 3+ inputs + set optimize='auto' without opt-einsum installed
- the tests will still pass if you have no opt-einsum

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -5135,6 +5135,18 @@ def c(t):
     inputs.append(SampleInput([c(H)], args=("i...->...",)))
     inputs.append(SampleInput([c(C), c(x)], args=('...ik, ...j -> ij',)))
 
+    # Test path kwarg
+    z = make_tensor((3, 2), device, dtype, requires_grad=requires_grad)
```
Comment (janeyx99) [inline torch/testing/_internal/common_methods_invocations.py:5139]:
```suggestion
    z = make_tensor((3, 2), device=device, dtype=dtype, requires_grad=requires_grad)
```

**Comment by soulitzer** [comment]
Just to bikeshed a bit on how the path is passed, maybe we could pass in strings to a "contraction_order_strategy" argument, e.g. "input-order", "opt-einsum", "maybe-opt-einsum", "path-arg" and assert that the "path" arg must be None unless the strategy is "path-arg" (these may not be the best names for the args/options)

**Comment by janeyx99** [comment]
> "contraction_order_strategy" argument, e.g. "input-order", "opt-einsum", "maybe-opt-einsum", "path-arg" and assert that the "path"

hahaha these are pretty creative--but I think this doesn't make the argument less confusing. I used to call this "optimize_path" but then realized that for jit, the C++ api and the python api have to match to some extent.

**Review by soulitzer — APPROVED** [review]
Nice!

File: torch/functional.py [hunk 2]
```diff
@@ -247,10 +248,14 @@ def einsum(*args: Any) -> Tensor:
...
-        thus, opt_einsum relies on different heuristics to achieve near-optimal results. If opt_einsum is not available,
-        the default order is to contract from left to right.
+        This function automatically uses opt_einsum (https://optimized-einsum.readthedocs.io/en/stable/) to speed up
+        computation or to consume less memory by optimizing contraction order. Note that finding _the_ optimal path is
+        an NP-hard problem, thus, opt_einsum relies on different heuristics to achieve near-optimal results. If the
+        opt_einsum package is not available, the default order is to contract from left to right.
+
+        Using opt_einsum to compute an optimal path does incur a constant amount of performance overhead. One can use
+        the `path` kwarg to specify an already optimized path by passing in a List or to simply bypass the
+        calculation by passing in `False`. See the `path` keyword argument below for more details.
```
Comment (soulitzer) [inline torch/functional.py:258]: We may want to update this too

Comment (janeyx99) [reply torch/functional.py:258]: good catch!

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -5135,6 +5135,18 @@ def c(t):
...
+    inputs.append(SampleInput([c(A), z], args=('ij,jk',), kwargs={'path': [(0, 1)]}))
+    inputs.append(SampleInput([c(A), c(D), c(C), c(E)],
+                  args=('jk,ikl,ijk,lm->mk',),
+                  kwargs={'path': [(0, 2), (0, 1), (0, 1)]}))
+    inputs.append(SampleInput([c(A), c(D), c(C), c(E)],
+                  args=('jk,ikl,ijk,lm->mk',),
+                  kwargs={'path': True}))
+    inputs.append(SampleInput([c(A), c(D), c(C), c(E)],
+                  args=('jk,ikl,ijk,lm->mk',),
+                  kwargs={'path': False}))
```
Comment (soulitzer) [inline torch/testing/_internal/common_methods_invocations.py:5149]: you forgot to update these

**Review by albanD — COMMENTED** [review]
The change sounds ok.
One question though is: given that numpy already has such an API (https://numpy.org/doc/stable/reference/generated/numpy.einsum.html) why don't we use the same thing? That will help with numpy compat and I'm sure be less confusing for a few users.

File: test/test_linalg.py [hunk 4]
```diff
@@ -3668,8 +3668,16 @@ def test_qr_error_cases(self, device, dtype):
...
         if np_args is None:
             np_args = [arg.cpu().numpy() if isinstance(arg, torch.Tensor) else arg for arg in args]
-        res = torch.einsum(*args)
+
         ref = np.einsum(*np_args)
+
+        # use path optimization; requires opt_einsum to be installed
+        if TEST_OPT_EINSUM:
+            res = torch.einsum(*args, path='use_opt_einsum')
+            self.assertEqual(torch.from_numpy(np.array(ref)), res)
```
[the remark this answers is not in the document]
Comment (lezcano) [reply test/test_linalg.py:3677]: Also, iirc, `assertEqual` can take a numpy array and a torch tensor.

**Comment by janeyx99** [comment]
> given that numpy already has such an API (https://numpy.org/doc/stable/reference/generated/numpy.einsum.html) why don't we use the same thing?

Do we want absolute parity here or just as much as would make sense? e.g., they have 2 different options for optimization ('greedy' and 'optimal' which don't quite match opt-einsum's options ('auto' and 'optimal'). I can introduce all three 'greedy', 'auto', 'optimal' OR we can just not have string inputs entirely.

Another point is numpy accepts a tuple `('einsum_path', list[...])`. We can support this or...we could just not.

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -13420,6 +13440,8 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                # there's a test `test_einsum` in `test_jit.py` to handle this case
                # AssertionError: JIT Test does not execute any logic
                DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit'),
+               # JIT runs into errors with einsum kwargs, see test_einsum in test_jit for more details
```
Comment (janeyx99) [inline torch/testing/_internal/common_methods_invocations.py:13443]: Causes test failures like https://github.com/pytorch/pytorch/actions/runs/3177210746/jobs/5177587394
```
======================================================================
ERROR [2.436s]: test_torchvision_models_maxvit_t (__main__.TestVisionTracing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_fx.py", line 4185, in run_test
    script = torch.jit.script(graph)
  File "/opt/conda/lib/python3.7/site-packages/torch/jit/_script.py", line 1287, in script
    obj, torch.jit._recursive.infer_methods_to_compile
  File "/opt/conda/lib/python3.7/site-packages/torch/jit/_recursive.py", line 476, in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
  File "/opt/conda/lib/python3.7/site-packages/torch/jit/_recursive.py", line 542, in create_script_module_impl
    create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
  File "/opt/conda/lib/python3.7/site-packages/torch/jit/_recursive.py", line 393, in create_methods_and_properties_from_stubs
    concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
RuntimeError: 
Arguments for call are not valid.
The following variants are available:
  
  aten::einsum(str equation, Tensor[] tensors, *, int[]? path=None) -> Tensor:
  Keyword argument optimize unknown.
  
  aten::einsum.sublist(Tensor a, ...) -> Tensor:
  Expected a value of type 'Tensor' for argument 'a' but instead found type 'str'.

The original call is:
  File "<eval_with_key>.25", line 59
    permute_3 = reshape_4.permute(0, 1, 3, 2, 4);  reshape_4 = None
    mul_2 = permute_2 * 0.125;  permute_2 = None
    einsum = torch.functional.einsum('B G H I D, B G H J D -> B G H I J', permute_1, mul_2, optimize = None);  permute_1 = mul_2 = None
             ~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    blocks_0_layers_0_layers_window_attention_attn_layer_1_relative_position_bias_table = getattr(getattr(getattr(self.blocks, "0").layers, "0").layers.window_attention.attn_layer, "1").relative_position_bias_table
    _tensor_constant0 = self._tensor_constant0
```

File: torch/functional.py [hunk 6]
```diff
@@ -197,7 +197,8 @@ def split(
     return tensor.split(split_size_or_sections, dim)
 
 
-def einsum(*args: Any) -> Tensor:
+def einsum(*args: Any,
+           optimize: Optional[Union[str, Sequence[int], Sequence[Tuple[int, int]]]] = None) -> Tensor:
     r"""einsum(equation, *operands) -> Tensor
```
Comment (albanD) [inline torch/functional.py:202]: This signature linee needs updating

**Review by albanD — APPROVED** [review]
Sounds good!

File: torch/functional.py (hunk 6 above)
Comment (janeyx99) [reply torch/functional.py:202]: o good catch!

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -5120,6 +5121,25 @@ def c(t):
...
     inputs.append(SampleInput([c(C), c(x)], args=('...ik, ...j -> ij',)))
 
+    # Test optimize kwarg
+    z = make_tensor((3, 2), device=device, dtype=dtype, requires_grad=requires_grad)
+    inputs.append(SampleInput([c(A), z], args=('ij,jk',), kwargs={'optimize': [(0, 1)]}))
+    inputs.append(SampleInput([c(A), c(D), c(C), c(E)],
+                  args=('jk,ikl,ijk,lm->mk',),
+                  kwargs={'optimize': [(0, 2), (0, 1), (0, 1)]}))
+    if TEST_OPT_EINSUM:
+        inputs.append(SampleInput([c(A), c(D), c(C), c(E)],
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5131]: fyi these SampleInputs could be updated to just yield each sample, and thanks to @peterbell10 the sample inputs can now be written "naturally" as regular inputs without dealing with the `args` or `kwargs` params, like:

```
yield SampleInput([c(A), c(D), c(C), c(E)], ('jk,ikl,ijk,lm->mk',), optimize='auto')
```

Comment (janeyx99) [reply torch/testing/_internal/common_methods_invocations.py:5131]: oh that is p cool!

**Comment by janeyx99** [comment]
This change has been superceded by #86219 and can be relanded later when we figure out JIT. Closing in the meantime

**Comment by albanD** [comment]
Could you open an issue to track landing this PR?

**Comment by janeyx99** [comment]
I’ll do one better—I’m going to make an issue tracking einsum perf improvement next steps in general, including this aspect.

**Pull request closed without merging**